spec/unit/policyfile_spec.rb

```diff
@@ -0,0 +1,145 @@
+      end
+    end
+  end
+
+  describe '#supermarket_source_url' do
+    let(:url) { 'https://supermarket.chef.io/api/v1/cookbooks/users' }
+
+    context 'when response not empty' do
+      it 'returns valid git repository url' do
+        supermarket_response = '{
+          "name": "users",
+          "maintainer": "chef",
+          "description": "Creates users from a databag search",
+          "category": "Other",
+          "latest_version": "https://supermarket.chef.io/api/v1/cookbooks/users/versions/5.3.1",
+          "external_url": "https://github.com/chef-cookbooks/users",
+          "source_url": "https://github.com/chef-cookbooks/users",
+          "issues_url": "https://github.com/chef-cookbooks/users/issues",
+          "average_rating": null,
+          "created_at": "2010-07-27T05:34:01.000Z",
+          "updated_at": "2017-12-15T18:08:26.990Z",
+          "up_for_adoption": null,
+          "deprecated": false,
+          "versions": [],
+          "metrics": {}
+        }'
+        stub_request(:get, url).to_return(status: 200, body: supermarket_response)
+        expect(changelog.supermarket_source_url(url)).to match(%r{^(http|https):\/\/.+\.git$})
+      end
+    end
+
+    context 'when cookbook does not exist' do
+      it 'raises an error' do
+        supermarket_response = '{ "error_messages":["Resource does not exist."],"error_code":"NOT_FOUND" }'
+        stub_request(:get, url).to_return(status: 404, body: supermarket_response)
+        expect { changelog.supermarket_source_url(url) }.to raise_error(RestClient::NotFound)
+      end
+    end
+  end
+
+  describe '#git_changelog' do
+    let(:git) { Git }
+
+    let(:git_repo) { double(Git::Base.new) }
+
+    let(:git_commit) do
+      double(
+        sha: 'e1b971a32f3a582766e4f62022ef7ed88e5eb8ba',
+        message: "Add test commit message\nThis line should not be shown"
+      )
+    end
+
+    context 'when given two tags' do
+      it 'generates a changelog between two tags' do
+        allow(changelog).to receive(:tag_format).and_return('v')
+        allow(git).to receive(:clone).and_return(git_repo)
+        allow(git_repo).to receive_message_chain(:log, :between)
+          .with('v1.0.0', 'v1.0.1')
+          .and_return([git_commit])
+
+        expect(changelog.git_changelog('https://url.example.git', '1.0.0', '1.0.1'))
+          .to eq('e1b971a Add test commit message')
+      end
+    end
+  end
+
+  describe '#tag_format' do
+    context 'when it receives a tag' do
+      let(:repo) { Git::Base }
+
+      it 'detects type for regular tag' do
+        allow(repo).to receive_message_chain(:tags, :last, :name).and_return('1.0.0')
+        expect(changelog.tag_format(repo)).to eq('')
+      end
+
+      it 'detects type for v-tag' do
+        allow(repo).to receive_message_chain(:tags, :last, :name).and_return('v1.0.0')
+        expect(changelog.tag_format(repo)).to eq('v')
+      end
+    end
+  end
+end
```

Comment:
Unfortunately, there is no test for generate_changelog which is the method you definitely need to test, this is your public contract.

Reply:
I've covered this method in the next commit. I also added tests for two other methods.

Reply:
Great!